### PR TITLE
Feature/purchasing events

### PR DESCRIPTION
**Pull request by bwhicks**
This PR addresses issues #130, #131, #132 with the following:

* Refactors `PartialDate` model logic as a mixin used in two models and easily extendable to any that we wish to have such logic in the future. 
* Refactors admin form validation for `PartialDate` as a `forms.ModelForm` subclass
* Add partial dates and footnotes to `Purchase`
* Enables `Purchase` editing in admin and adds `Purchase` inline to `Account` change form.

### CODE REVIEW

File: mep/accounts/admin.py [hunk 1]
```diff
@@ -196,6 +203,77 @@ class ReimbursementAdmin(admin.ModelAdmin):
...
+                        'address data.'),
+        }
+        widgets = {
+            'account': AUTOCOMPLETE['account'],
+            'item': AUTOCOMPLETE['item']
+        }
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # override start date label to just date, since reimbursements
```
Comment (rlskoeser) [inline mep/accounts/admin.py]: tiny quibble, reimbursements should be purchases here

File: mep/accounts/admin.py [hunk 2]
```diff
@@ -49,6 +49,13 @@
     ),
 }
 
+
+class OpenFootnoteInline(FootnoteInline):
+    '''Customize footnote inline for borrowing and purchase events.'''
+    classes = ('grp-collapse', )  # grapelli collapsible, but not closed
+    extra = 0
```
Comment (rlskoeser) [inline mep/accounts/admin.py]: I set extra to 0 for borrowing events so it would show the one created on import without an extra one, but for data entry on purchases we expect them to be adding footnotes. I think maybe we should set this to 1 instead of 0.

File: mep/accounts/admin.py [hunk 3]
```diff
@@ -255,7 +334,8 @@ class AccountAdmin(admin.ModelAdmin):
                      'address__location__name',
                      'address__location__country__name', 'persons__name')
     fields = ('persons', 'card')
-    inlines = [AccountAddressInline, SubscriptionInline, ReimbursementInline]
+    inlines = [AccountAddressInline, SubscriptionInline, PurchaseInline,
```
Comment (rlskoeser) [inline mep/accounts/admin.py]: It doesn't make sense to me to include purchases here, since borrows aren't included - an account with lots of purchases could make this form slow and overwhelming. If we do decide to keep it, I think it should be listed last.

File: mep/accounts/tests/test_accounts_admin.py [hunk 4]
```diff
@@ -65,7 +65,13 @@ def test_clean(self):
...
         assert form.cleaned_data['end_date'] == form_data['end_date']
 
-class TestBorrowAdminForm(TestCase):
+
+class TestPartialDateMixin(TestCase):
+
+    # Letting this test stand for the partial date mixin,
+    # since that's the only functionality being checked here,
+    # and it is common to both sets + mixin needs a concrete
+    # case to test against
```
Comment (rlskoeser) [inline mep/accounts/tests/test_accounts_admin.py]: This comment as written doesn't make sense to me. Maybe reword to just document that it's testing partial date mixin via borrow, since it needs a concrete case?

File: mep/accounts/models.py [hunk 5]
```diff
@@ -619,18 +600,85 @@ def date_range(self):
...
+        self.end_date_precision = self.start_date_precision
+        self.end_date = self.start_date
+        super().save(*args, **kwargs)
+
+    def validate_unique(self, *args, **kwargs):
+        '''Validation check to prevent duplicate purchase events from happening.
+        Differs from
+        :class:`~mep.accounts.models.Reimbursement.validate_unique` by also
+        allowing for multiple purchases of different items per day.
+        '''
```
Comment (rlskoeser) [inline mep/accounts/models.py]: Possible worth noting that we can't use `unique_together` because of multi-table inheritance. (I know it's noted on the other, but I forgot and had to look at it to see why we did this.)

File: mep/accounts/admin.py (hunk 3 above)
Comment (bwhicks) [reply mep/accounts/admin.py]: I was seeing `Purchase` as analogous to `Reimbursement`,  since they're very samey, but I see they're closer to `Borrow` based on this and our brief discussion yesterday. Happy to pull it out.

File: mep/accounts/tests/test_accounts_admin.py (hunk 4 above)
Comment (bwhicks) [reply mep/accounts/tests/test_accounts_admin.py]: Revising accordingly.
